src/core/server/saved_objects/validation/validator.ts

```diff
@@ -0,0 +1,129 @@
+ *
+ * See {@link SavedObjectsValidationFunction} for custom validation.
+ *
+ * @public
+ */
+export type SavedObjectsValidationSpec<A extends SavedObjectAttributes> =
+  | ObjectType
+  | SavedObjectsValidationFunction<A>;
+
+/**
+ * A map of {@link SavedObjectsValidationSpec | validation specs} to be used for a given type.
+ * The map's keys must be valid semver versions.
+ *
+ * Any time you change the schema of a {@link SavedObjectsType}, you should add a new entry
+ * to this map for the Kibana version the change was introduced in.
+ *
+ * @example
+ * ```typescript
+ * const validationMap: SavedObjectValidationMap = {
+ *   '1.0.0': schema.object({
+ *     foo: schema.string(),
+ *   }),
+ *   '1.1.0': schema.object({
+ *     foo: schema.oneOf([schema.string(), schema.boolean()]),
+ *   }),
+ *   '2.1.0': (data) => {
+ *     if (typeof data.bar !== 'string') {
+ *       throw new Error(`[bar]: expected value of type [string] but got [${typeof data.bar}]`);
+ *     }
+ *     if (typeof data.foo !== 'string' && typeof data.foo !== 'boolean') {
+ *       throw new Error(`[foo]: expected value of type [string,boolean] but got [${typeof data.foo}]`);
+ *     }
+ *   }
+ * }
+ * ```
+ *
+ * @public
+ */
+export interface SavedObjectsValidationMap<
+  A extends SavedObjectAttributes = SavedObjectAttributes
+> {
+  [version: string]: SavedObjectsValidationSpec<A>;
+}
+
+/**
+ * Helper class that takes a {@link SavedObjectsValidationMap} and runs validations for a
+ * given type based on the provided Kibana version.
+ *
+ * @internal
+ */
+export class SavedObjectsTypeValidator<A extends SavedObjectAttributes = SavedObjectAttributes> {
+  private readonly type: string;
+  private readonly validationMap: SavedObjectsValidationMap<A>;
+
+  constructor({
+    type,
+    validationMap,
+  }: {
+    type: string;
+    validationMap: SavedObjectsValidationMap<A> | (() => SavedObjectsValidationMap<A>);
+  }) {
+    this.type = type;
+    this.validationMap = typeof validationMap === 'function' ? validationMap() : validationMap;
+  }
+
+  public validate(kibanaVersion: string, data: A): void {
+    const validationRule = this.validationMap[kibanaVersion];
+    if (!validationRule) {
+      return; // no matching validation rule could be found; proceed without validating
+    }
+
+    if (isConfigSchema(validationRule)) {
+      validationRule.validate(data, {});
+    } else if (isValidationFunction(validationRule)) {
+      this.validateFunction(data, validationRule);
+    } else {
+      throw new ValidationError(
+        new SavedObjectsValidationError(
+          `The ${kibanaVersion} validation for saved object of type [${this.type}] is malformed.`
+        )
+      );
+    }
+  }
+
+  private validateFunction(data: A, validateFn: SavedObjectsValidationFunction) {
+    try {
+      validateFn({ ...data }); // shallow clone to prevent mutation
```

Comment:
As mentioned in my other comment, maybe this should be future-proofed?
```suggestion
      validateFn({ attributes: { ...data } }); // shallow clone to prevent mutation
```

Reply:
shallow clone does not really prevent mutation, e.g

```ts
const data = {
   foo: {
      bar: 12
   }
};

const shallow = { ...data };

shallow.foo.bar = 7;
data.foo.bar // <== 7
```

(and having attributes with nested properties is quite common)

So as I said on a previous comment, I think we should either KISS and do nothing for now, or do it properly via a proper deepCopy or another mutation-prevention pattern.

Reply:
> Yeah I know, I was trying to hit a middle ground between the perf hit of a deep clone while still providing some protection against mutation. 

The problem with this approach that is implicit. `validateFn({ attributes: Object.freeze(data) } });` will throw an error, so Kibana devs can see the mutations are explicitly forbidden.